enhancements/machine-api/remediation-history.md

```diff
@@ -0,0 +1,217 @@
+---
+title: remediation-history
+authors:
+  - "@slintes"
+reviewers:
+  - "@beekhof"
+approvers:
+  - "@JoelSpeed"
+  - "@michaelgugino"
+  - "@enxebre"
+creation-date: 2020-12-15
+last-updated: 2021-02-08
+status: implementable
+---
+
+# Remediation history
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Operational readiness criteria is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+Record and show limited remediation history.
+
+## Motivation
+
+With machine health checks administrators can configure automatic remediation of node problems. There is no good
+overview though for when and why nodes were remediated. Recording and showing a history of remediations would allow
+administrators to review when and why machines were remediated, and to discover patterns like e.g. machineA always
+reboots at 2am on Wednesdays.
+
+### Goals
+
+- Record a limited remediation history.
+- Show remediation history in the UI.
+
+### Non-Goals
+
+- Record an *unlimited* history of remediations.
+- Collect or provide information on *why* a node has a certain node condition which triggered remediation.
+
+## Proposal
+
+The machine health check holds the configuration for automatic remediation of nodes. We want to extend the status of
+machine health checks to contain the last x remediation events, with information about which node is affected, which
+condition triggered the remediation, and the timestamps of
+- when the triggering condition was detected
+- when remediation was started
+- when the node was successfully fenced (deleted / powered off)
+
+This information can be displayed in the UI in a table on a machine healthcheck details page.
+
+MHC type enhancement:
+
+```go
+
+// MachineHealthCheckStatus defines the observed state of MachineHealthCheck
+type MachineHealthCheckStatus struct {
+
+	[...]
+
+	// History of remediations triggered by this machine health check
+	RemediationHistory []Remediation `json:"remediationHistory,omitempty"`
+}
+
+// Remediation tracks a remediation triggered by this machine health check
+type Remediation struct {
+	// a reference to the target machine or node
```

Comment:
Which of these will it be, machine or node? I would have assumed we would only ever point to one of these kinds? (Probably the machine?)

Reply:
It depends on what exactly triggers remediation. Machines in failed phase or without a node can do so, then the machine will be referenced. When it's a node condition, it will be the node.

Reply:
Right ok, that makes more sense. I wonder if it's worth adding to the comment here why it's machine or node, just that context you added would be good IMO.

Reply:
done